eng/scripts/MavenPackaging.ps1

```diff
@@ -153,3 +153,66 @@ function Get-FilteredMavenPackageDetails([string]$ArtifactDirectory, [string]$Gr
 
   return $filteredPackageDetails
 }
+
+# Compare the contents of SHA hash files for a local repository to the same
+# files in Maven Central or Azure DevOps. Returns $false if none of the hash
+# files exist in the remote, $true if all hashes exist and match the local
+# hash, otherwise throws.
+function Test-ReleasedPackage([string]$RepositoryUrl, [MavenPackageDetail]$PackageDetail, [string]$BearerToken) {
+  if ($RepositoryUrl -match "^https://pkgs.dev.azure.com/azure-sdk/\b(internal|public)\b/*") {
+    if(!$BearerToken) {
+      throw "BearerToken required for Azure DevOps package feeds"
+    }
+    
+    $baseUrl = "https://pkgs.dev.azure.com/azure-sdk/internal/_packaging/azure-sdk-for-java-pr/maven/v1"
+    $algorithm = "sha256"
+    $headers = @{ Authorization="BEARER $BearerToken" }
+  }
+  elseif ($RepositoryUrl -match "^https://oss.sonatype.org/service/local/staging/deploy/maven2") {
+    $baseUrl = "https://repo1.maven.org/maven2"
+    $algorithm = "sha1"
+  }
+  else {
+    throw "Repository URL must be either an Azure Artifacts feed, or a SonaType Nexus feed."
+  }
+
+  $packageUrl = "$baseUrl/$($PackageDetail.GroupId.Replace('.', '/'))/$($PackageDetail.ArtifactID)/$($PackageDetail.Version)"
+  
+  # Count the number of remote hashes found
+  $remoteCount = 0
+
+  # Count the number of remote hashes that match their local hash
+  $matchCount = 0
+
+  foreach($artifact in $PackageDetail.AssociatedArtifacts) {
+    $localFileName = $artifact.File.Name;
+    $remoteHashUrl = "$packageUrl/$localFileName.$algorithm"
+
+    Write-Information "Comparing local and remote hashes for $localFileName"
+    Write-Information "  Getting remote hash"
+    $response = Invoke-WebRequest -Method GET -Uri $remoteHashUrl -Headers $headers -SkipHttpErrorCheck
+
+    if($response.StatusCode -ge 200 -and $response.StatusCode -lt 300) {
+      $remoteCount++
+      $remoteHash = $response.Content
+
+      Write-Information "  Getting local hash"
+      $localPath = $artifact.File.FullName
+      $localHash = Get-FileHash -Path $localPath -Algorithm $algorithm | Select-Object -ExpandProperty 'Hash'
+      
+      if($remoteHash -eq $localHash) {
+        $matchCount++
+        Write-Information "  Remote $remoteHash == Local $localHash"
+      } else {
+        Write-Information "  Remote $remoteHash != Local $localHash"
+      }
+    }
+    else {
+      Write-Information "  Unable to retrieve remote hash for $localFileName. Http reponse code: $($response.StatusCode)"
+    }
```

Comment:
If we do plan to check all should we fail at the first missing or miss-matched hash? It doesn't seem like we should continue to try the others if one fails.

Reply:
Misses are expected when nothing was deployed, so we don't want to fail on first missing remote hash.  We could fail on first mismatch and on first miss when remoteCount > 0.

Reply:
I'm not sure I follow. If nothing was deployed we will fail on the first one and why wouldn't we fail fast and not worry about checking the others because we already know things are broken? Or is this to help debug partial uploads of some files but not others? If we end up in that state I not even sure how we recover.

Reply:
It's really cheap to check everything.  So if we're bad, just check the rest, log it all, then throw.

We can't short circuit out at the first missing file.   If everything's missing, that's an expected state, so just return `$false`.   If only one file is missing, that's unexpected so `throw`.

Reply:
If nothing's there in the remote, it's not a failure, it's just a lack of remote content.

Reply:
OK I think I had a different expectation of this function which was that it would essentially return true if the package was there and valid otherwise return false which means try to publish again. However I now understand what you are going for and I'm fine with throwing in the case where we don't have some but not all the artifacts or if some of them are invalid.